### ISSUE TITLE
Conversion to full matrix

### ISSUE THREAD

**Issue by flay** [nservant/HiC-Pro#40]
It would be great to have an option to convert the triplet sparse format to full matrix format that can eventually be compatible with downstream analysis tools such as TAD callers from Dixon et al and Crane et al.

**Comment by nservant** [comment]
Hi,
I just committed a first version of the sparseToDense utility to convert sparse matrix into dense format.
usage: sparseToDense.py [-h] [--lengths LENGTHS] [--bins BINS]
                        [--output OUTPUT]
                        filename
If --bins is defined with the BED file of genomic bins, it should create the input files required by the Bing Ren TADs caller. Otherwise, it will just write the output matrix.
This is first beta version. Any comments are welcome.
Hope it helps
Nicolas

**Comment by flay** [comment]
Hi, Thank you for this. I was wondering  what is the input file format for --lengths here? I thought it was chrom.sizes similar to what you get from UCSC, but I am getting an error for loading the --lengths file.

**Comment by nservant** [comment]
Hi. Thanks for the test. I will look at that asap and keep you informed.

**Comment by nservant** [comment]
Sorry for the delay. I just released a new version.
This script requires the last version of the iced package 2.2, that I just pull with the master of HiCPro or which is available as an independent python module.

usage: sparseToDense.py [-h] [-b BINS] [-d] [-o OUTPUT] filename

positional arguments:
  filename

optional arguments:
  -h, --help            show this help message and exit
  -b BINS, --bins BINS  BED file with bins coordinates. If provided the
                        chromosome lengths are used to define the output
                        matrix size
  -d, --di              If specified the output matrix is formatted for Dixon
                        et al. TADs calling. In this case --bins is required
  -o OUTPUT, --output OUTPUT
                        Output filename

Hope it helps
Nicolas

**Comment by suhnrhie** [comment]
Hi, 
I tried sparseToDense.py, but I get error messages. Did I put the input file wrong? or is this bug? I had no problem installing iced package (used HiC-Pro_2.7.7/scripts/src/iced-0.2.2 and HiC-Pro_2.7.7/scripts/src/ice_mod).

This is the command I used.
python $PATH1/HiC-Pro_2.7.7/bin/utils/sparseToDense.py -b $PATH2/HiC-Pro/HiC.run/OUTPUT/hic_results/matrix/rawdata/raw/20000/rawdata_20000_ord.bed -d -o HiC.20kb.iced.hicpro.di $PATH2/HiC-Pro/HiC.run/OUTPUT/hic_results/matrix/rawdata/iced/20000/rawdata_20000_iced.matrix

The is the error message I got.
Traceback (most recent call last):
  File "sparseToDense.py", line 34, in <module>
    counts = io.load_counts(args.filename, lengths=lengths)
  File "/Python-2.7.11/venv/lib/python2.7/site-packages/iced/io/_io_else.py", line 30, in load_counts
    counts = sparse.coo_matrix((X[:, 2], (X[:, 0], X[:, 1])), shape=shape)
  File "/Python-2.7.11/venv/lib/python2.7/site-packages/scipy/sparse/coo.py", line 182, in __init__
    self._check()
  File "/Python-2.7.11/venv/lib/python2.7/site-packages/scipy/sparse/coo.py", line 238, in _check
    raise ValueError('column index exceeds matrix dimensions')
ValueError: column index exceeds matrix dimensions

**Comment by nservant** [comment]
Hi,
Looks strange. It seems that there is something wrong between the BED file and the matrix file.
Does it work if your remove the -b and -d option ?

**Comment by suhnrhie** [comment]
When I remove the -b and -d options, I get different errors.

python sparseToDense.py -o HiC.20kb.iced.hicpro.di  $PATH2/HiC-Pro/HiC.run/OUTPUT/hic_results/matrix/rawdata/iced/20000/rawdata_20000_iced.matrix
                                 
Traceback (most recent call last):
  File "sparseToDense.py", line 36, in <module>
    counts = counts + counts.T
ValueError: operands could not be broadcast together with shapes (155129,155131) (155131,155129)

**Comment by nservant** [comment]
Hi,
It looks like your matrix is not square ! could you please check wich parameter you put in the MATRIX_FORMAT  parameter ? On which organism are you working ? Thank you.

**Comment by suhnrhie** [comment]
I just used the HiC-Pro output matrix from $PATH2/HiC-Pro/HiC.run/OUTPUT/hic_results/matrix/rawdata/iced/20000/rawdata_20000_iced.matrix

141133748 lines for rawdata_20000_iced.matrix

head rawdata_20000_iced.matrix

1	1	15
1	5	1
1	6	1
1	42	1
1	388	1
1	1378	1
1	1697	1
1	2190	1
1	2920	1
1	4955	1

155130 lines for rawdata_20000_ord.bed

head rawdata_20000_ord.bed
chr1	0	20000	1
chr1	20000	40000	2
chr1	40000	60000	3
chr1	60000	80000	4
chr1	80000	100000	5
chr1	100000	120000	6
chr1	120000	140000	7
chr1	140000	160000	8
chr1	160000	180000	9
chr1	180000	200000	10

Should I use different files? Which file I should use from HiC-Pro pipeline output files? I used human (hg19).

Thanks,

**Comment by suhnrhie** [comment]
This is the command I used to run HiC-Pro

HiC-Pro -i $PATH2/HiC-Pro/HiC -o OUTPUT -c config-hicpro.txt

I put my HiC fastq files in the folder, $PATH2/HiC-Pro/HiC/rawdata

After the run, all of HiC-Pro processed files are found in  $PATH2/HiC-Pro/HiC.run/OUTPUT/

**Comment by nservant** [comment]
I think everything is fine in the way you run HiCPro.
In the BED file you have 155130 line so in the sparseToDense script the matrix should be of size 155130x155130 which is not the case (you have 155129 x 155131)
Did you build the maps at lower resolution too ? Does it work with other resolution ?
I guess you are using the latest version of the iced package, 0.2.2 ? If not, could you please update it.
I will need the reproduice the bug, would you agree to share these files with me ?
Thanks
Nicolas

**Comment by suhnrhie** [comment]
I am not sure how come I get the error >,< I am using the iced package 0.2.2 (the one you uploaded in HiC-Pro_2.7.7/scripts/src/iced-0.2.2)
 
[]$ python
Python 2.7.6 (default, Dec 17 2013, 17:06:01)
[GCC 4.4.7 20120313 (Red Hat 4.4.7-4)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import iced
>>> iced.__version__
'0.2.2-git'

I tried in another machine freshly, and interestingly, I get different errors
python /HiCPro/HiC-Pro_2.7.7/bin/utils/sparseToDense.py -b rawdata_20000_ord.bed -d -o test.matrix.txt rawdata_20000_iced.matrix

Traceback (most recent call last):
  File "/HiCPro/HiC-Pro_2.7.7/bin/utils/sparseToDense.py", line 35, in <module>
    counts = counts.toarray()
  File "/.local/lib/python2.7/site-packages/scipy/sparse/coo.py", line 274, in toarray
    B = self._process_toarray_args(order, out)
  File "/.local/lib/python2.7/site-packages/scipy/sparse/base.py", line 793, in _process_toarray_args
    return np.zeros(self.shape, dtype=self.dtype, order=order)
MemoryError

I guess it could be due to memory error.

When I tried with 1000kb ones, it worked ;)

python /HiCPro/HiC-Pro_2.7.7/bin/utils/sparseToDense.py -b rawdata_1000000_ord.bed -d -o test.matrix.txt rawdata_1000000_iced.matrix

However, test.matrix.txt from this run give errors in the step calling HMM using domain caller (Dixon et al, 2012)
matlab < HMM_calls.m > dumpfile
Error using chol
Matrix must be positive definite.

Error in gaussLogprob (line 52)
  R    = chol(Sigma);

Error in mixGaussInferLatent (line 17)
  logPz(:, k) = logMix(k) + gaussLogprob(mu(:, k), Sigma(:, :, k), X);

Error in mixGaussFit>estep (line 53)
[weights, ll] = mixGaussInferLatent(model, data);

Error in emAlgo (line 60)
    [ess, ll] = estep(model, data);

Error in mixGaussFit (line 26)
[model, loglikHist] = emAlgo(model, data, initFn, @estep, @mstep , ...

I think it's due to the fact that the test.matrix.txt from HiC-Pro (after running sparseToDense.py) have a different distribution from the example test dataset that Bing Ren group uploaded in http://chromosome.sdsc.edu/mouse/hi-c/download.html - I was able to call TAD using example test dataset from this link using domain caller. 

I also tested using the example test data you uploaded https://zerkalo.curie.fr/partage/HiC-Pro/HiCPro_testdata.tar.gz

I was able to make test.matrix.txt for 1000000 bin size one, but got same errors as above at the step calling HMM in domain caller. 

I also tried to use directionalityIndex function in HiTC R package after making HTClist object by importing HiC-Pro matrix, but same errors found. 

In summary, the matrix file made from sparseToDense.py doesn't work for the domain caller (Dixon et al, 2012). 

If you would like, I can share my data, but I think using your HiCPro_testdata, you should be able to test why HiC-Pro output files do not work for the domain caller.

**Comment by suhnrhie** [comment]
By the way, I used upper for matrix_format. 
MATRIX_FORMAT = upper

What does this parameter mean? Does the result differ if I use complete, asis, or lower? I cannot find any detailed description of this parameter.

**Comment by nservant** [comment]
I just change the delimiter of sparseToDense output and push the new code.
For the Memory Error. I reproduced it once, but now it's fine. 
The amount of RAM require to store a dense matrix is huge. 
We will think about less memory intensive way to getting the inputs for the TADs calling.
For instance, one idea would be to write only cis maps for each chromosome.

**Comment by suhnrhie** [comment]
Of course, I changed the matrix file from sparseToDense to DI using DI_from_matrix.pl. There was no problem making DI_from_matrix.pl, but the DI file from your matrix file doesn't work with HMM. Have you tried to run matlab HMM using the DI file? I think matrix file from HiC-Pro has different distribution. Could you please check if you can run HMM using DI from matrix file of HiC-Pro raw and iced?

**Comment by nservant** [comment]
ok I took me some times, but I just did it and it worked.
Actually, using this code is not that easy. So here is what I did ;

0- I'm in the iced/1000000 folder, and I used the test dataset

0bis - I'm not a Matlab expert, but it seems that the libraries or modules need to be loaded from the current folder. So I created a link to the require modules from the domains_call sotware itself !!!
`ln -s Apps/domains_call/domaincall_software/required_modules/ require_modules`

1- Convert to dense matrix. It creates the file  
`Apps/HiC-Pro_2.7.7/bin/utils/sparseToDense.py dixon_2M_1000000_iced.matrix -b ../../raw/1000000/dixon_2M_1000000_abs.bed -d`

2- Create the DI file
Apps/domains_call/domaincall_software/perl_scripts/DI_from_matrix.pl `dixon_2M_1000000_iced_dense.matrix 1000000 2000000 Apps/HiC-Pro_2.7.7/annotation/chrom_hg19.sizes > dixon_2M_1000000_iced_dense.di`

3- Then, we need a couple of additional changes because the domains caller does not like chrX, Y and M
`grep -v "M" dixon_2M_1000000_iced_dense.di | grep -v "Y" | sed -e 's/^X/23/g' > dixon_2M_1000000_iced_dense_v2.di
`
4- Copy the file HMM_calls.m in the current folder and manually edit it (!!!) to set the input, ie. the v2.di file and the output ...

5- run Matlab
`nice /bioinfo/opt/Matlab/bin/matlab < HMM_calls_iced_test.m`

And it should works :(
I know that using the domains calling's code is tricky. But regarding the conversion to dense matrix and HiC-pro, I think it is fine ...
I can still have a look for high resolution to improve the memory usage
Btw, I will moe this discussion on the HiC-Pro forum, as I think others could be interested ;
https://groups.google.com/forum/#!forum/hic-pro

**Comment by suhnrhie** [comment]
Hi Nicolas, 

Do you have a new script written which doesn't require the large memory?

I also have questions about the sparseToDense.py function. Does this script just rearrange the values in abs.bed and matrix from HiC-Pro output into one big matrix? Or this script does additional thing?
I wrote the script by myself to make a one big matrix and created DI file from this file by running DI_from_matrix.pl, but somehow, I get all positive values for DI. 
I also used HiTC R package to make a matrix and created DI from the matrix by running DI_from_matrix.pl, and I get all positive values for DI.
Interestingly, I used HiTC R package to make a matrix and called DI using directionalityIndex function, and I get all negative values for DI.

Dixon et al HMM requires to have DI from negative to positive values. I guess the reason that I get error I posted above "Error using chol Matrix must be positive definite." in the step 5-run matlab could be due to wrong calculation of DI. 

I have done these analyses with published data as well (Dixon et al ESC HiC data and GSM1902607 HiC data) and I get errors. So, this is definitely not due to HMM or HiC dataset itself. Maybe, HiC-Pro is not properly normalizing data or creating matrix/DI? Could you please help?

Thanks,

**Comment by nservant** [comment]
Hi,
We did not update the sparseToDense.py script since version 2.7.7.
Actually, the memory error appears when you manage in RAM the dense genome-wide matrix.
I'm not sure to see how I can improve this part ...
The sparseToDense script only performs a matrix transformation, from sparse to dense matrix. There is nothing more ...
So if I understand correctly, when you load the data, and run the DI score, you only have positive values ? And you think that this is why the HMM calling is crashing ?
I just checked on one of my project that I analysed with HiCPro, run the DI score and I have negative and positive values ... so I do not think that this is due to normalization ... but is will require additional checks.
Could you try with the Dixon raw contact maps please ? or could you send it to me (with your config file) ?
And I'm sorry, but I will be out for a couple of weeks ... and should be back at the end of the month
Nicolas

**Comment by suhnrhie** [comment]
Hi Nicolas,

I think I found out the cause of errors getting all positive and negative values for DI scores.
The script I wrote by myself (I still cannot run sparseToDense.py in my machine due to memory) was not making a mirror matrix, it had data only in upper triangular of the matrix. Now, I fixed to make a full matrix, and I get pos and neg values for DI scores.

When I used HiTC R package to make a matrix, I also realized that the matrix from HiTC didn't have a full matrix, it had data only in upper triangular of the matrix. That was why I think I got all positive values for DI when I run DI_from_matrix.pl from the matrix from HiTC R package.

However, I am not sure why I get all negative values for DI when I used HiTC R package using directionalityIndex. Could you please check your HiTC R package functions, so I can use HiTC for visualization/DI as well? For you, I am posting the R scripts I used with HiTC R package here. 

library("HiTC")
x=importC("../rawdata_40000_iced.matrix", xgi.bed="../rawdata_40000_ord.bed")
R=extractRegion(x$chr1chr1, chr="chr1", from=1, to=249250621)
di<-directionalityIndex(R)
ord=read.delim(paste("rawdata_40000_ord.chr1.cut.bed",sep="."), header=F)
RM=as.matrix(intdata(R))
RF=cbind(ord[,c(1:3)], RM)
#RF is the matrix from HiTC
ord$DI=di
ord$V1=paste("1")
ord$DI[is.na(ord$DI)]<-0
ord$DI[is.nan(ord$DI)]<-0
FIN.DI=ord[,c(1,2,3,4)]
# FIN.DI is the DI file from HiTC

By the way, is there any way that I can use sparseToDense.py script with only one chromosome? As I mentioned, I still cannot run the sparseToDense.py due to memory, and I don't know how I can use this script with only one chromosome.

Thanks,

**Comment by nservant** [comment]
Hi 
Sorry for the delay, and thank you very much for your feedback.
Yes I confirm that by default the data are stored as a sparse symmetric Matrix in HiTC.
I will correct this bug as soon as possible.

And for sparseToDense, this is also something with is in the TODO list !!
Thanks
Nicolas

**Comment by nservant** [comment]
You can simply change the line
    xdata <- intdata(x)
by 
    xdata <- as.matrix(intdata(x))

in the directionalityIndex function. This should solve the issue.
Nicolas

**Comment by suhnrhie** [comment]
Well.. I changed xdata as you suggested, but I still get all negative values from directionalityIndex. 

> directionalityIndex
function (x, winup = 2e+06, windown = 2e+06)
{
    stopifnot(isBinned(x))
    xdata <-as.matrix(intdata(x))
    rx <- ranges(x_intervals(x))
    upranges <- IRanges(start(rx) - winup + 1, start(rx))
    A <- rep(NA, length(upranges))
    dwranges <- IRanges(end(rx), end(rx) + windown - 1)
    B <- rep(NA, length(upranges))
    overup <- findOverlaps(rx, upranges, type = "within")
    overdown <- findOverlaps(rx, dwranges, type = "within")
    for (i in 1:length(rx)) {
        idx <- which(subjectHits(overup) == i)
        if (length(idx) > 0) {
            idx.rx <- queryHits(overup)[idx]
            A[i] <- sum(xdata[idx.rx, i], na.rm = TRUE)
        }
        idx <- which(subjectHits(overdown) == i)
        if (length(idx) > 0) {
            idx.rx <- queryHits(overdown)[idx]
            B[i] <- sum(xdata[idx.rx, i], na.rm = TRUE)
        }
    }
    E <- (A + B)/2
    DI <- ((B - A)/abs(B - A)) * (((A - E)^2)/E + ((B - E)^2)/E)
    DI
}
> R=extractRegion(x$chr1chr1, chr="chr1", from=1, to=249250621)
> di<-directionalityIndex(R)
summary(di)
> summary(di)
    Min.  1st Qu.   Median     Mean  3rd Qu.     Max.     NA's
-5213.00 -2144.00 -1812.00 -1841.00 -1496.00   -38.43      578

If you save xdata as txt and visualize in excel, you will see that the file has only upper part of data.

**Comment by nservant** [comment]
Hi
Sorry, actually it was
`    xdata <- as.matrix(intdata(forceSymmetric(x)))`

The forceSymmetric method converts the count data to a sparse symmetric object.
Then the conversion using as.matrix should recreate a dense matrix.
I did a test, and it fix the issue with my data.

Could you first check whether 
`as.matrix(intdata(forceSymmetric(R)))[1:10, 1:10]`
gives you a dense matrix. If so, it should be fine.

Nicolas

**Comment by suhnrhie** [comment]
Hi, 

The forceSymmetric function works. I can see a dense matrix. For now, I'd better stick with HiTC after making a sparse matrix using HiC-Pro.

Thanks,

**Comment by Shaedeycool** [comment]
Hi,

I have tried to use sparsetodense.py on output from HiC-Pro:

HiC-Pro/bin/utils/sparseToDense.py -b sample1/raw/1000000/sample1_1000000_abs.bed sample1/iced/1000000/sample1_1000000_iced.matrix

I receive the error:

/Users/shaedeycool/miniconda3/envs/dataframes/lib/python3.10/site-packages/iced/normalization/_ca_utils.py:8: UserWarning: The API of this module is likely to change. Use only for testing purposes
  warnings.warn(
/miniconda3/envs/dataframes/lib/python3.10/site-packages/iced/io/_io_pandas.py:56: UserWarning: Attempting to guess whether counts are 0 or 1 based
  warnings.warn(
Traceback (most recent call last):
  File "/Volumes/SADE_PG/HiC-Pro/bin/utils/sparseToDense.py", line 92, in <module>
    counts = io.load_counts(args.filename, lengths=lengths)
  File "/Users/shaedeycool/miniconda3/envs/dataframes/lib/python3.10/site-packages/iced/io/_io_pandas.py", line 72, in load_counts
    counts = sparse.coo_matrix((data, (row, col)), shape=shape)
  File "/Users/shaedeycool/miniconda3/envs/dataframes/lib/python3.10/site-packages/scipy/sparse/_coo.py", line 196, in __init__
    self._check()
  File "/Users/shaedeycool/miniconda3/envs/dataframes/lib/python3.10/site-packages/scipy/sparse/_coo.py", line 283, in _check
    raise ValueError('row index exceeds matrix dimensions')
ValueError: row index exceeds matrix dimensions

the matrix file has 2589589 lines however the bed file only has 24 lines. Using hg38.

Help in fixing this problem would be greatly appreciated.
